### PR TITLE
Feature: chain store

### PR DESCRIPTION
**Pull request by masterjedy**
### Description of the Change
ChainStore is storage backend for chain manager.
It handles bookkeeping, 
enables receiving block, uses block validator to validate it before adding;
enables sampling randomness from chain;
allows to see which block has been finalized most recently.

### Benefits
ChainStore enables implementing ChainManager.

### Possible Drawbacks 
Specs are incomplete, convenience methods need to be implemented later.
Tests to go.

### CODE REVIEW

File: cmake/dependencies.cmake [hunk 1]
```diff
@@ -49,3 +49,4 @@ find_package(cppcodec CONFIG REQUIRED)
 # https://hunter.readthedocs.io/en/stable/packages/pkg/Protobuf.html
 hunter_add_package(Protobuf)
 find_package(Protobuf CONFIG REQUIRED)
+
```
Comment (Mingela) [inline cmake/dependencies.cmake]: :suspect:

File: core/blockchain/impl/weight_calculator_impl.cpp [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+ */
+
+#include "blockchain/impl/weight_calculator_impl.hpp"
+
+namespace fc::blockchain::weight {
+  using primitives::BigInt;
+
+  outcome::result<BigInt> WeightCalculatorImpl::calculateWeight(
+      const Tipset &tipset) {
+    return 1;  // this value is fake, make proper implementation
```
Comment (Mingela) [inline core/blockchain/impl/weight_calculator_impl.cpp]: mention TODO for that as well, FIL-155 I guess

File: core/primitives/chain/chain.cpp [hunk 3]
```diff
@@ -0,0 +1,10 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "primitives/chain/chain.hpp"
+
+namespace primitives::chain {
+
```
Comment (Mingela) [inline core/primitives/chain/chain.cpp]: is it supposed to be empty?

File: core/storage/chain/chain_data_store.hpp [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+namespace fc::storage::blockchain {
+
+  class ChainDataStore {
+   public:
+    virtual ~ChainDataStore() = default;
+    /** @brief get value by key */
+    virtual outcome::result<std::string> get(const DatastoreKey &key) const = 0;
+    /** @brief set key -> value */
+    virtual outcome::result<void> set(const DatastoreKey &key,
+                                      std::string_view value) = 0;
```
Comment (Mingela) [inline core/storage/chain/chain_data_store.hpp]: will there be any string validation, it's quite strange if it possible to put any string value to a chain data store

File: core/storage/chain/chain_data_store.hpp [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+ */
+
+#ifndef CPP_FILECOIN_CORE_STORAGE_CHAIN_CHAIN_DATA_STORE_HPP
+#define CPP_FILECOIN_CORE_STORAGE_CHAIN_CHAIN_DATA_STORE_HPP
+
+#include "storage/chain/datastore_key.hpp"
+
+namespace fc::storage::blockchain {
+
+  class ChainDataStore {
```
Comment (Mingela) [inline core/storage/chain/chain_data_store.hpp]: probably some comment is needed why do we need this

File: core/storage/chain/datastore_key.hpp [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+     */
+    static DatastoreKey makeFromString(std::string_view value) noexcept;
+  };
+
+  enum class DatastoreKeyError {
+    INVALID_DATASTORE_KEY = 1,  /// invalid data used for creating datastore key
+  };
+
+  /** @brief cbor-encodes DatastoreKey instance */
+  template <class Stream,
```
Comment (turuslan) [inline core/storage/chain/datastore_key.hpp]: `CBOR_ENCODE(DatastoreKey, key) {` (from "codec/cbor/streams_annotation.hpp")

File: core/primitives/chain/chain.hpp [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+namespace fc::primitives::chain {
+  /**
+   * @struct Chain is a sequence of tipsets linked together
+   */
+  struct Chain {
+    tipset::Tipset head_tipset;
+
+    outcome::result<std::reference_wrapper<const tipset::Tipset>>
+    getTipsetAtEpoch(ChainEpoch epoch) const;
```
[the remark this answers is not in the document]
Comment (Mingela) [reply core/primitives/chain/chain.hpp]: UPD. Yea, they say Chain Manager should track all branches/forks as separate objects and clear them on finalization

File: core/storage/chain/chain_data_store.hpp (hunk 4 above)
Comment (masterjedy) [reply core/storage/chain/chain_data_store.hpp]: I don't think it is strange. It looks normal for me. What kind of validation you suggest?

File: core/primitives/chain/chain.cpp (hunk 3 above)
Comment (masterjedy) [reply core/primitives/chain/chain.cpp]: yes, for time being

File: core/primitives/chain/chain.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (masterjedy) [reply core/primitives/chain/chain.hpp]: this interface is going to be implemented in chain manager PR, currently only chain store is fulfilled

File: core/primitives/chain/chain.cpp (hunk 3 above)
Comment (masterjedy) [reply core/primitives/chain/chain.cpp]: It's only interface for chain manager (fil-85), currently only chain store is implemented

**Pull request merged**